File: backend/parse_pdf_to_db.js
const fs = require('fs');
const pdf = require('pdf-parse');
const { PrismaClient } = require('@prisma/client');
const prisma = new PrismaClient();

async function run() {
    console.log("Parsing PDF...");
    const dataBuffer = fs.readFileSync('d:/setu tss ui ux/grants and scheme/77a5d438-b239-4ec1-a539-d08804635b0b.pdf');
    const data = await pdf(dataBuffer);
    const text = data.text;

    // We split by numbered headings like "1. Name", "2. Name"
    // Since the OCR shows "1. Central Sector Scheme...", we can use regex
    const splits = text.split(/\n\d+\.\s+/);
    splits.shift(); // remove everything before the first scheme

    console.log("Found " + splits.length + " schemes. Clearing old grants...");
    await prisma.grant.deleteMany({});

    for (const chunk of splits) {
        let title = chunk.split('\n')[0].trim();
        let ministry = (chunk.match(/Ministry:\s*(.+)/) || [])[1] || "";
        let sectors = (chunk.match(/Sectors:\s*(.+)/) || [])[1] || "";
        let brief = (chunk.match(/Brief:\s*([\s\S]+?)(?:Eligibility Criteria:|Benefits:|Benefit Tags:)/) || [])[1] || "";
        let eligibility = (chunk.match(/Eligibility Criteria:\s*([\s\S]+?)(?:Benefits:|Benefit Tags:|Quantum\/Size)/) || [])[1] || "";
        let benefits = (chunk.match(/Benefits:\s*([\s\S]+?)(?:Benefit Tags:|Quantum\/Size|Tenure:|Notes:|Apply:)/) || [])[1] || "";
        let benefitTags = (chunk.match(/Benefit Tags:\s*(.+)/) || [])[1] || "";
        let quantum = (chunk.match(/Quantum\/Size of Fund:\s*([\s\S]+?)(?:Tenure:|Notes:|Apply:)/) || [])[1] || "";
        let tenure = (chunk.match(/Tenure:\s*(.+)/) || [])[1] || "";
        let link = (chunk.match(/Apply:\s*(.+)/) || [])[1] || "";

        const clean = (s) => s.trim().replace(/\n/g, ' ');

        await prisma.grant.create({
            data: {
                title: clean(title),
                ministry: clean(ministry),
                sectors: clean(sectors),
                description: clean(brief),
                eligibility: clean(eligibility),
                benefits: clean(benefits),
                benefitTags: clean(benefitTags),
                amount: clean(quantum),
                tenure: clean(tenure),
                link: clean(link),
                is_active: true
            }
        });
    }

    console.log("Successfully seeded grants!");
    process.exit(0);
}

run().catch(console.error);

File: trash/parse_pdf_to_db.py
import fitz
import re
import psycopg2

print("Parsing PDF...")
doc = fitz.open("../grants and scheme/77a5d438-b239-4ec1-a539-d08804635b0b.pdf")
text = ""
for page in doc:
    text += page.get_text()

splits = re.split(r"\n\d+\.\s+", text)
splits.pop(0)

print(f"Found {len(splits)} schemes.")

# Update this with your DB connection string
DATABASE_URL = "postgres://postgres:postgres@localhost:5432/db"

try:
    conn = psycopg2.connect(DATABASE_URL)
    cur = conn.cursor()
    cur.execute("DELETE FROM grants")
    
    for chunk in splits:
        title = chunk.split("\n")[0].strip()
        ministry = re.search(r"Ministry:\s*(.+)", chunk)
        sectors = re.search(r"Sectors:\s*(.+)", chunk)
        brief = re.search(r"Brief:\s*([\s\S]+?)(?:Eligibility Criteria:|Benefits:|Benefit Tags:)", chunk)
        eligibility = re.search(r"Eligibility Criteria:\s*([\s\S]+?)(?:Benefits:|Benefit Tags:|Quantum/Size)", chunk)
        benefits = re.search(r"Benefits:\s*([\s\S]+?)(?:Benefit Tags:|Quantum/Size|Tenure:|Notes:|Apply:)", chunk)
        benefitTags = re.search(r"Benefit Tags:\s*(.+)", chunk)
        quantum = re.search(r"Quantum/Size of Fund:\s*([\s\S]+?)(?:Tenure:|Notes:|Apply:)", chunk)
        tenure = re.search(r"Tenure:\s*(.+)", chunk)
        link = re.search(r"Apply:\s*(.+)", chunk)
        
        cur.execute("""
            INSERT INTO grants (id, title, ministry, sectors, description, eligibility, benefits, "benefitTags", amount, tenure, link, is_active, created_at)
            VALUES (gen_random_uuid(), %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, true, NOW())
        """, (
            title,
            ministry.group(1).strip() if ministry else "",
            sectors.group(1).strip() if sectors else "",
            brief.group(1).strip().replace("\n", " ") if brief else "",
            eligibility.group(1).strip().replace("\n", " ") if eligibility else "",
            benefits.group(1).strip().replace("\n", " ") if benefits else "",
            benefitTags.group(1).strip() if benefitTags else "",
            quantum.group(1).strip().replace("\n", " ") if quantum else "",
            tenure.group(1).strip() if tenure else "",
            link.group(1).strip() if link else ""
        ))
        
    conn.commit()
    print("Database seeded successfully!")
except Exception as e:
    print("Error:", e)
finally:
    if 'conn' in locals():
        conn.close()

